### DISCUSSION TITLE
Custom RGB random colors for Effects - Rog Strix 18

### DISCUSSION THREAD

**Discussion by Dawhack** [seerge/g-helper#2807] [Ideas]
I wonder if it's possible to have the custom color of RGB in the effects like "stars" etc. instead of just one color or one set of color because i know it's possible to do it on Armory Crate.
Is there any way to do it on g-helper? thank you very much

**Comment by Dawhack** [comment]
What i mean is that in armory crate in most effects you can choose to random RGB or a custom set of RGB insteaf of just one color like in g-helper, is there any way to make it show on g-helper too?

**Comment by seerge** [comment]
@Dawhack hello, currently app lists all modes and settings supported on a firmware level by your keyboard (with exception if `Ambient`). I.e. the ones that can run automatically w/o any app being involved since they are set.

Based on what i know modes like `Stars` - don't support custom colors by default. AC may "imitate" them by literally sending custom colors for each individual button few times a second (but that costs of a lot of resources to run).

Can you check if custom colors are available in AC if you **disable** "Aura Sync" there ?

**Reply by Dawhack** [reply]
Hello @seerge  thank you very much for your answer.
I can find the "random" rgb effects even without Aura Sync, I can select "random" on the colors and RGB will show random colors on the effects and that looks very good.
It works also on other effects like "Ripple" (looks very good), "Laser", "Stars", "Highlight" (looks good) and probably others.
Those looks great and if it's possible to choose it on Ghelper it would be amazing.
Thank you again for all you work.

**Reply by seerge** [reply]
@Dawhack hello, ok.

Can you make sure that Aura Sync is completely disabled in Armoury. And then post screenshots of possible settings  from AC  for each effect ?

**Reply by Dawhack** [reply]
@
![a1](https://github.com/seerge/g-helper/assets/87214127/730c8a3a-6bcd-40aa-b299-dfe4b600878a)
![a2](https://github.com/seerge/g-helper/assets/87214127/38500297-a035-43d7-a93b-db83cd8ed43f)
![a3](https://github.com/seerge/g-helper/assets/87214127/0b69905d-750e-4de7-a4d3-2b33c957340d)
![a4](https://github.com/seerge/g-helper/assets/87214127/5b03913d-5d67-4112-a500-a710b9c6a161)
![672f7b2650fd1e1b67b4410490dd57f8](https://github.com/seerge/g-helper/assets/87214127/ea2e8cdd-29e6-45d0-add4-ca21486fbed8)


Sorry for late response, yea definitely disabled and you can still select direction and random rgb color on effects.
I did not made a screenshot for every one effect but if you need i would do that no problem.

Thank you again.

**Reply by Dawhack** [reply]
On the .gif is not visible but Aura Sync is disabled.
My laptop is Asus Rog Strix 18 2023

**Reply by seerge** [reply]
@Dawhack ok, great. 

But single color for each effect should be already possible with G-Helper as well. At least it works for any effect on my G14 and as far as i know on other devices.

Are you missing only "random" color ? How does this random work in AC ? Does it change with time, or just sets one color and stays like that after?

**Reply by Dawhack** [reply]
Thank you again for your reply and your hard work, yes it is possible to heave a single color for each effect in G-helper, but the problem is that you can't choose to "randozime" colors like in AC, with "randomize" it will show differents colors on the effect and not just one color.

The option "random" makes the effect show not just one color but differents colors (random RGB). That looks really amazing and i miss that option on the great G-helper. 
If you need any more information i'm 100% avaiable to give you everything. Thank you again.

**Reply by seerge** [reply]
@Dawhack if those "random" colors are handled by firmware it must be some specific command that AC sends. You can try to record it with a special tool.

----

Posting manual from other issue

Install https://www.hhdsoftware.com/device-monitoring-studio (will require reboot)

1. Launch Monitoring Studio, pick Asus keyboard device (one with product id from the list) 
https://github.com/seerge/g-helper/blob/a33089921edbac9a49335f31c7c2cc1c117e5fb0/app/USB/AsusHid.cs#L13

2. Double click on it
![Screenshot 2024-01-20 163125](https://github.com/seerge/g-helper/assets/5920850/e9a28802-08cb-423c-812b-7b526b051a5f)

3. Add ``Packet View`` to Procesing and click ``Start``
![Screenshot 2024-01-20 163136](https://github.com/seerge/g-helper/assets/5920850/354340f0-2fcb-473b-a49f-2e01b25d00a9)

4. Go to Armoury and set one of those modes + random color there.

5. Check if Monitoring Studio catching a lot of data. If NOT - you probably have selected wrong Asus device on step 2 (so need to re-do it)

6. In Monitoring Studio select ``Tools`` -> ``Save to Log``, and save data
![Screenshot 2024-01-20 163247](https://github.com/seerge/g-helper/assets/5920850/ad480dcf-05bc-43a7-9a80-2984fb68e8a8)

Upload this data here.

**Reply by seerge** [reply]
@Dawhack but before that, can you quickly try this build and check if it accidentally sets random color ?

[GHelper.zip](https://github.com/user-attachments/files/16117417/GHelper.zip)

**Reply by Dawhack** [reply]
Sure, i'm trying this build first

**Reply by Dawhack** [reply]
![sss](https://github.com/seerge/g-helper/assets/87214127/bce62ecb-35bb-4640-96a3-0d9de7e91ee1)
unfortunately doesn't work

**Reply by seerge** [reply]
@Dawhack ok, and this ? color selector will look same, just after changing any of the modes (that support random in AC) keyboard itself should go random

[GHelper.zip](https://github.com/user-attachments/files/16117445/GHelper.zip)

**Reply by Dawhack** [reply]
It works! i was about to run Device Monitorning Studio and couldn't find my keyboard there.. but this works!! wow!!!! amazing

**Reply by seerge** [reply]
@Dawhack ok, great. Can you mark modes that support random colors ? 

Full list for Stirx models

P.S. No need to record anything then

```
            { AuraMode.AuraStatic, Properties.Strings.AuraStatic },
            { AuraMode.AuraBreathe, Properties.Strings.AuraBreathe },
            { AuraMode.AuraColorCycle, Properties.Strings.AuraColorCycle },
            { AuraMode.AuraRainbow, Properties.Strings.AuraRainbow },
            { AuraMode.Star, "Star" },
            { AuraMode.Rain, "Rain" },
            { AuraMode.Highlight, "Highlight" },
            { AuraMode.Laser, "Laser" },
            { AuraMode.Ripple, "Ripple" },
            { AuraMode.AuraStrobe, Properties.Strings.AuraStrobe},
            { AuraMode.Comet, "Comet" },
            { AuraMode.Flash, "Flash" },
            { AuraMode.HEATMAP, "Heatmap"},
            { AuraMode.AMBIENT, "Ambient"},
```

**Reply by seerge** [reply]
@Dawhack and main question, i'm not sure where to add this random flag / checkbox in the UI

**Reply by Dawhack** [reply]
So the effects that support "random colors" are:
Rain
Highlight
Laser
Ripple

I'm not fully sure but i think that's it

**Reply by Dawhack** [reply]
I'm checking now to be 100% sure.
Also for the Checkbox i'm not sure also! I think it's a good idea to have it as "default" and if someone want to change it he can still do it from the color box!

**Reply by Dawhack** [reply]
Yes that's it, those 4 effects have the randomize rgb effect! and it looks very good.

**Reply by seerge** [reply]
@Dawhack ok, I can propose to use pure "Black" color as a toggle for random colors here (so no extra UI changes needed). Cause Black is not used here anyway.

I.e. if you want to set random color - set it to Black, otherwise - use any other color and it will work as well directly

Try this build
[GHelper.zip](https://github.com/user-attachments/files/16117494/GHelper.zip)

**Reply by Dawhack** [reply]
That is a very great idea! In this build unfortunately the black toggle doesn't work as "random colors"
It just make everything "black" (off).

**Reply by seerge** [reply]
@Dawhack ok, can you try this build instead ?

[GHelper.zip](https://github.com/user-attachments/files/16117525/GHelper.zip)

**Reply by Dawhack** [reply]
Tried right now and everything works, those 4  effects with "black" set as color make the "random color" appear!
You're amazing.

**Reply by seerge** [reply]
@Dawhack ok, great. Then I guess we can consider this as resolved (I will add add a notice about Black color meaning somewhere in the FAQ and release notes)

**Reply by Dawhack** [reply]
Thanks man, it's solved ;)

**Reply by fluent2** [reply]
Hey there! So I've been actively interested in the Rain setting for awhile now. The good news is that the white color can be replaced by a custom color (the far left bottom corner!!) So this is great. But it seems there's no way to change the default Red color huh? That's a bit sad as we are halfway there now, lol, just missing that option. Great work tho, thanks!

**Reply by seerge** [reply]
@fluent2 does `Rain` setting have 2 colors in AC ? (with AuraSync **disabled**)

If so try to select `Breathing` mode (it will let you select 2 colors), set both of them and switch again to `Rain` and check if colors work

**Reply by fluent2** [reply]
> @fluent2 does `Rain` setting have 2 colors in AC ? (with AuraSync **disabled**)
> 
> If so try to select `Breathing` mode (it will let you select 2 colors), set both of them and switch again to `Rain` and check if colors work

Hey seerge, thanks for the attempt to help! So I tried this and when switching back to Rain, the one color only is used, and the Red remains. Not sure what else to try here, but if there's any way I can test anything, just let me know!

Edit - Now that I test it more, seems that nothing changes the default colors, and I can't seem to reproduce changing the white to lime green. 🤷

**Reply by seerge** [reply]
@fluent2 does Rain setting have 2 colors in AC ? (with AuraSync disabled)

**Reply by fluent2** [reply]
> @fluent2 does Rain setting have 2 colors in AC ? (with AuraSync disabled)

Nope, just one color. And that color doesn't seem to work for me. Also, just to make sure, how do I disable AuraSync just in case??

**Reply by seerge** [reply]
@fluent2 ok, but then it looks like `Rain` doesn't support a second customizable color at all ? Then most likely it's not possible to change it.

AuraSync is somewhere in AC settings (check @Dawhack 's screenshots).
